app/vmctl/mimir/lazyreader.go

```
@@ -0,0 +1,175 @@
+		chunk, err := lbr.fetchFile(blockChunkPath)
+		if err != nil {
+			log.Printf("failed to fetch chunk file: %q: %s", chunkName, err)
+			return err
+		}
+		if err := lbr.writeFile(temp, blockChunkPath, chunk); err != nil {
+			log.Printf("failed to write chunk file: %q: %s", chunkName, err)
+			return err
+		}
+	}
+
+	// Set postingDecoder to nil because
+	// If it is nil then a default decoder is used, compatible with Prometheus v2.
+	pb, err := tsdb.OpenBlock(nil, temp, nil, nil)
+	if err != nil {
+		return fmt.Errorf("failed to open block %q: %s", lbr.ID, err)
+	}
+	lbr.reader = pb
+	blockID := lbr.ID.String()
+	blockPath := filepath.Join(temp, blockID)
+	if err := os.RemoveAll(blockPath); err != nil {
+		log.Printf("failed to remove temp dir: %s", err)
+		return fmt.Errorf("failed to remove temp dir: %s", err)
+	}
+	return nil
+}
+
+// Index returns an IndexReader over the block's data.
+func (lbr *LazyBlockReader) Index() (tsdb.IndexReader, error) {
+	if err := lbr.initialize(); err != nil {
+		return nil, err
+	}
+	return lbr.reader.Index()
+}
+
+// Chunks returns a ChunkReader over the block's data.
+func (lbr *LazyBlockReader) Chunks() (tsdb.ChunkReader, error) {
+	if err := lbr.initialize(); err != nil {
+		return nil, err
+	}
+	return lbr.reader.Chunks()
+}
+
+// Tombstones returns a tombstones.Reader over the block's deleted data.
+func (lbr *LazyBlockReader) Tombstones() (tombstones.Reader, error) {
+	if err := lbr.initialize(); err != nil {
+		return nil, err
+	}
+	return lbr.reader.Tombstones()
+}
+
+// Meta provides meta information about the block reader.
+func (lbr *LazyBlockReader) Meta() tsdb.BlockMeta {
+	if err := lbr.initialize(); err != nil {
+		log.Printf("error get Block Meta: %s; return empty block", err)
+		return tsdb.BlockMeta{}
+	}
+	return lbr.reader.Meta()
+}
+
+// Size returns the number of bytes that the block takes up on disk.
+func (lbr *LazyBlockReader) Size() int64 {
+	if err := lbr.initialize(); err != nil {
```

Again, err is ignored.

---

updated method realisation